lib/puppet/resource/type.rb

```diff
@@ -563,10 +563,22 @@ def create_params_struct
   #
   def lookup_external_default_for(param, scope)
     return unless type == :hostclass
+
     qname = "#{name}::#{param}"
-    in_global = lambda { lookup_with_databinding(qname, scope) }
-    in_env = lambda { lookup_in_environment(qname, scope) }
-    in_module = lambda { lookup_in_module(qname, scope) }
+
+    if Puppet[:hiera_advanced_parameter_bindings]
+      opt_name = "lookupoptions::#{qname}"
+      in_global = lambda { lookup_with_databinding(opt_name, scope) }
+      in_env = lambda { lookup_in_environment(opt_name, scope) }
+      in_module = lambda { lookup_in_module(opt_name, scope) }
+      merge_strategy = lookup_search(in_global, in_env, in_module)
```

Comment:
The above can be performed with a call to `Puppet::Pops::Lookup.lookup` where it is also possible to to a type assertion of the result. (See the puppet/functions/lookup.rb for type signatures for the various options).